rpc-server/src/dispatchers/blockchain.rs

```diff
@@ -468,55 +525,57 @@ impl BlockchainInterface for BlockchainDispatcher {
         include_stakers: Option<bool>,
-    ) -> Result<Validator, Error> {
+    ) -> Result<BlockchainState<Validator>, Error> {
         let blockchain = self.blockchain.read();
 
-        let accounts_tree = &blockchain.state().accounts.tree;
-        let db_txn = blockchain.read_transaction();
-        let validator = StakingContract::get_validator(accounts_tree, &db_txn, &address);
-
-        if validator.is_none() {
-            return Err(Error::ValidatorNotFound(address));
-        }
-
-        let mut stakers = None;
-
-        if include_stakers == Some(true) {
-            let staker_addresses =
-                StakingContract::get_validator_stakers(accounts_tree, &db_txn, &address);
-
-            let mut stakers_map = HashMap::new();
-
-            for address in staker_addresses {
-                let staker = StakingContract::get_staker(accounts_tree, &db_txn, &address).unwrap();
-                if !staker.balance.is_zero() {
-                    stakers_map.insert(address, staker.balance);
-                }
-            }
-
-            stakers = Some(stakers_map);
-        }
-
-        Ok(Validator::from_validator(&validator.unwrap(), stakers))
+        get_validator_by_address(blockchain.deref(), &address, include_stakers)
     }
 
     /// Tries to fetch a staker information given its address.
-    async fn get_staker_by_address(&mut self, address: Address) -> Result<Staker, Error> {
+    async fn get_staker_by_address(
+        &mut self,
+        address: Address,
+    ) -> Result<BlockchainState<Staker>, Error> {
         let blockchain = self.blockchain.read();
 
         let accounts_tree = &blockchain.state().accounts.tree;
         let db_txn = blockchain.read_transaction();
         let staker = StakingContract::get_staker(accounts_tree, &db_txn, &address);
 
+        let block_number = blockchain.block_number();
+        let block_hash = blockchain.head_hash();
+
         match staker {
-            Some(s) => Ok(Staker::from_staker(&s)),
+            Some(s) => Ok(BlockchainState::from_value(
+                block_number,
+                block_hash,
+                Staker::from_staker(&s),
+            )),
             None => Err(Error::StakerNotFound(address)),
         }
     }
 
-    /// Subscribes to blockchain events.
+    /// Subscribes to new block events (retrieves the full block).
     #[stream]
-    async fn head_subscribe(&mut self) -> Result<BoxStream<'static, Blake2bHash>, Error> {
+    async fn head_subscribe(
+        &mut self,
+        include_transactions: Option<bool>,
+    ) -> Result<BoxStream<'static, Result<Block, Blake2bHash>>, Error> {
+        let blockchain = Arc::clone(&self.blockchain);
+        let stream = self.head_hash_subscribe().await?;
+
+        Ok(stream
+            .map(move |hash| {
+                let blockchain_rg = blockchain.read();
+                get_block_by_hash(blockchain_rg.deref(), &hash, include_transactions)
+                    .map_err(|_| hash)
```

Comment:
Does this, and the result type `Result<Block, Blake2bHash>` mean that this stream can _either_ return a block object _or_ a hash string?

Reply:
You are correct. So this function uses the head_hash_subscribe to have a stream of new block hashs and then fetches the block using that hash. In case there is a new block committed, the new hash is received over the stream. If the block was reverted in the meantime, the get_block_by_hash fails. In that case, we are sending the hash as a return value. 

The reason for this behavior is that serde is not able to serialize something like Result<Block, Error>. The alternative would be to filter out the event altogether. That could work too, I'm not seeing any issues with it.